File: ui/runtime/src/protocol/messages/device.ts
import { ResultResponse } from './common';
import { LocalizedText } from '../../localization/localized-text';

export type DeviceClientType = 'web-client' | 'admin-ui' | 'native' | 'provider' | 'unknown';

export type DeviceFormFactor = 'phone' | 'tablet' | 'desktop' | 'unknown';

export interface Device {
  id: string;
  name: string;
  nameIsCustom: boolean;
  proposedName?: string;
  clientType: DeviceClientType;
  formFactor: DeviceFormFactor;
  platform?: string;
  browser?: string;
  appVersion?: string;
  online: boolean;
  connectionCount: number;
  hasActiveSession: boolean;
  lastSeenAt: string;
  createdAt: string;
  startupProfileId?: string;
  startupProfileName?: string;

  providerId?: string;
  providerName?: LocalizedText;
  providerDeviceId?: string;
  model?: string;
  manufacturer?: string;
  layoutReference?: string;
}

export interface GetDevicesResponse {
  devices: Device[];
}

export interface RenameDeviceRequest {
  name: string;
}

export interface RenameDeviceResponse extends ResultResponse {
  device?: Device;
}

export interface SetDeviceStartupProfileRequest {
  profileId: string | null;
}

export interface SetDeviceStartupProfileResponse extends ResultResponse {
  device?: Device;
}

export interface OpenProfileOnDeviceRequest {
  profileId: string;
}

export interface OpenProfileOnDeviceResponse extends ResultResponse {}

export interface LogoutDeviceResponse extends ResultResponse {}

export interface RemoveDeviceResponse extends ResultResponse {}

export interface DeviceChangedEvent {
  device: Device;
}

export interface DeviceRemovedEvent {
  deviceId: string;
}

export interface DeviceSessionRevokedEvent {}

export interface DeviceLoginInfo {
  deviceId?: string;
  deviceSecret?: string;
  clientType?: DeviceClientType;
  proposedName?: string;
  platform?: string;
  browser?: string;
  formFactor?: DeviceFormFactor;
  appVersion?: string;
}

export interface DeviceCredential {
  deviceId: string;
  deviceSecret?: string;
  startupProfileId?: string;
}

File: host/src/MacroDeckHost.Application/Ui/Transport/Messages/Devices/Device.cs
using MacroDeck.Localization;

namespace MacroDeckHost.Application.Ui.Transport.Messages.Devices;

public class Device
{
	public string Id { get; set; } = string.Empty;
	public string Name { get; set; } = string.Empty;
	public bool NameIsCustom { get; set; }
	public string? ProposedName { get; set; }

	public string ClientType { get; set; } = "unknown";

	public string FormFactor { get; set; } = "unknown";

	public string? Platform { get; set; }
	public string? Browser { get; set; }
	public string? AppVersion { get; set; }

	public bool Online { get; set; }
	public int ConnectionCount { get; set; }
	public bool HasActiveSession { get; set; }

	public string? StartupProfileId { get; set; }

	public string? StartupProfileName { get; set; }

	public DateTime LastSeenAt { get; set; }

	public DateTime CreatedAt { get; set; }

	/// <summary>Id of the integration or plugin providing this device; absent for a client device.</summary>
	public string? ProviderId { get; set; }

	/// <summary>Display name of the providing integration or plugin.</summary>
	public LocalizedText ProviderName { get; set; }

	public string? ProviderDeviceId { get; set; }

	public string? Model { get; set; }

	public string? Manufacturer { get; set; }

	public string? LayoutReference { get; set; }
}
